### PR TITLE
Fix travis build instability with MySQL and IBM DB2

### PR DESCRIPTION
**Pull request by mmucklo**
|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | no

#### Summary

Fixes the build instability seen with MySQL and IBM.

### CODE REVIEW

**Comment by mmucklo** [comment]
This adds two new exception subclasses to properly propagate and distinguish connection exceptions versus regular or driver exceptions.

This is necessary as the MySQL error 2006 (MySQL server has gone away) could happen after a successful connection, and probably should be distinguished from a pure connection error that happens during an attempt to connect.

**Comment by mmucklo** [comment]
@morozov - This is ready for review.

There's still some occasionally build failings due to VM / Network stability issues, which is hard for me to fix, but barring that, this should fix the general flakiness of the MySQL builds (due to unhandled error conditions), and updates appveyor to the only php version that's presently available on chocolate (which started failing over the weekend).

The IBM issue started happening last week or so, and the appveyor builds started failing over the weekend.

The ocular update is not strictly necessary, but was just one of my steps along the way in troubleshooting.

Mainly the two MySQL error conditions that weren't being caught are now caught. The mysql driver would seem to rotate through these errors and pick one to fail on fairly regularly.

The connection going away error (2006) can happen normally, so I extended the classes to try and distinguish it happening during the time of connection so that the DBAL ConnectionException is known to be thrown.

**Comment by morozov** [comment]
Which specific failure(s) does this patch fix?

**Comment by mmucklo** [comment]
@morozov 

TLDR; Please skip to the itemized list below.

Without these fixes, the build in the master branch will not work as far as my experience is concerned (notice the number of failed builds in the current list of PRs). For PR #3645 I spent way too long just trying to get a full build to run and all tests to pass. A portion of it was chasing down these issues with the build.

Incidentally I've spent **a lot** of time on this and the other PR (probably hours upon hours over many days). I'm not saying that you should accept things carte blanche, by no means, but hopefully I can work with you with whatever concerns you have as it would be such that the effort doesn't go to waste.

1. **MySQL** -
_(consistently flaky)_

Fixes ~25% chance of falling back to DriverException instead of ConnectionException leading to failed builds as the DBAL ConnectionException test fails (as it sees DriverException instead of CnonectionException).

Apparently trying to connect to a non-existent host can cause the mysqli / PDO MySQL drivers to crash in different ways, probably due to timings of the network stack during the run, but possibly just by some other race condition.

2. **IBM DB2** -
_(broken some time last few weeks or so)_

  apt update command failing, causing the build to consistently crash:
  [https://travis-ci.community/t/then-sudo-apt-get-update-failed-public-key-is-not-available-no-pubkey-6b05f25d762e3157-in-ubuntu-xenial/1728](https://travis-ci.community/t/then-sudo-apt-get-update-failed-public-key-is-not-available-no-pubkey-6b05f25d762e3157-in-ubuntu-xenial/1728)

3. **appveyor** -
_(broke some time last week or so)_

PHP 7.2 no longer available on chocolatey (which is what appveyor builds use to install php).
[https://chocolatey.org/packages?q=php&sortOrder=package-title&page=2&prerelease=False&moderatorQueue=False](https://chocolatey.org/packages?q=php&sortOrder=package-title&page=2&prerelease=False&moderatorQueue=False)

All appveyor builds were crashing as a result.

4. **scrutinizer - ocular version**
_(scrutinizer test coverage uploads were not working a few days ago)_

Updated to the latest scrutinizer ocular. Not a fix per se, but scrutinizer was bombing on test coverage import - see:

[https://github.com/scrutinizer-ci/ocular/issues/40#issuecomment-520241401](https://github.com/scrutinizer-ci/ocular/issues/40#issuecomment-520241401)

**Comment by mmucklo** [comment]
Closing this in favor of the 3 separated-out PRs:

#3653 
#3654 
#3656

**Comment by mmucklo** [comment]
@morozov Please see the 3 new PRs opened which address the issues separately:

#3653
#3654
#3656
